cmd/scollector/collectors/google_cloudbilling.go

```diff
@@ -0,0 +1,234 @@
+package collectors
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"golang.org/x/oauth2/google"
+
+	"google.golang.org/api/storage/v1"
+
+	"encoding/base64"
+
+	"encoding/json"
+
+	"strings"
+
+	"math"
+
+	"bosun.org/cmd/scollector/conf"
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+	"bosun.org/slog"
+)
+
+var gcbConf = conf.GoogleCloudBilling{}
+
+const (
+	gcbHoursInDay = 24
+	gcbUsageDesc  = "Usage of consuming given Google Cloud services"
+	gcbCostDesc   = "Cost of consuming given Google Cloud services"
+)
+
+// GoogleCloudBilling instantiates the GoogleCloudBilling collector
+func GoogleCloudBilling(bucketName, authBase64 string) error {
+	if bucketName != "" {
+		gcbConf = conf.GoogleCloudBilling{
+			BucketName: bucketName,
+			AuthBase64: authBase64,
+		}
+		collectors = append(collectors, &IntervalCollector{
+			F:        c_googlecloudbilling,
+			Interval: 1 * time.Hour,
+		})
+	}
+
+	return nil
+}
+
+func c_googlecloudbilling() (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	//Create a connection to the Google Cloud services
+	gcClient, err := createGCStorageClient()
+	if err != nil {
+		return nil, err
+	}
+	//Get the list of available files from the storage bucket
+	reports, err := gcClient.Objects.List(gcbConf.BucketName).Do()
+	if err != nil {
+		return nil, err
+	}
+	//We want to process bills from the last 4 days. This is because Google Cloud seems to have significant
+	//lag when generating the bill exports. They can be 2 or more days old by the time the bill is put into the
+	//storage bucket.
+	dateStart := time.Now().Add(-96 * time.Hour)
+	dateRangeStart := time.Date(dateStart.Year(), dateStart.Month(), dateStart.Day(), 0, 0, 0, 0, time.UTC)
```

Comment:
I couldn't get time.Truncate to do what I wanted here, because I don't want to actually round the time to an interval of 4 days, I just want "now minus 4 days" (or 96 hours, because screw timezone changes)